meta-lmp-base/classes/lmp.bbclass

```diff
@@ -30,6 +50,10 @@ IMAGE_CMD:ota:append () {
 		${OTA_SYSROOT}/ostree/deploy/${OSTREE_OSNAME}/var/sota/import/installed_versions \
 			${GARAGE_TARGET_NAME}-${target_version}
 	fi
+
+	if [ "${APP_PRELOAD_WITHIN_OE_BUILD}" = "1" ]; then
+		preload_apps
```

Comment:
@ricardosalveti Do you think it's possible to run this command as a root user? `su root -c preload_apps`.

Reply:
This command runs in `IMAGE_CMD:ota:append` when building the image, for that bitbake makes use of the tool `pseudo`  that work like `fakeroot` with some additional functionality. This to say that the `preload_apps` is already running with root privileges.

Reply:
For some reason `dockerd` doesn't think so. I just simply added  `dockerd &` to the preload-apps.sh and it complains that it doesn't have root permissions.

When I tried to run `dockerd-rootless` it complained that it cannot be started by a root user :).

I'll send a link to the error later on.

Reply:
I think this is another issue, running `dockerd &` is not possible at build time because because at build time the host architecture is the one in aws `amd64` and the target can be different as we use cross-compilation.

bitbake as the concept of native application (used during the build for run on the host and not on the target) but running a daemon is not supported afaik.

Reply:
I understand it, but this is very unlikely the reason because I am not running containers by means of `dockerd`, I am just running an executable (dockerd) built for amd64 on amd64 host.

This is the error message

```
| Preloading Apps of the given Target: intel-corei7-64-lmp-1367
| time="2022-05-31T17:40:08.869842422Z" level=info msg="Starting up"
| dockerd needs to be started with root privileges. To run dockerd in rootless mode as an unprivileged user, see https://docs.docker.com/go/rootless/
```

Reply:
https://ci.foundries.io/projects/msul-dev01/lmp/builds/1367/intel-corei7-64/artifacts/console.log

Reply:
This is the error message for the rootless dockerd

```
| |--
| Preloading Apps of the given Target: intel-corei7-64-lmp-1361
| + [ -w /home/builder/.docker/run ]
| + [ -d /home/builder ]
| + rootlesskit=
| + command -v docker-rootlesskit
| + command -v rootlesskit
| + rootlesskit=rootlesskit
| + break
| + [ -z rootlesskit ]
| + :
| + :
| + : builtin
| + : auto
| + : auto
| + net=
| + mtu=
| + [ -z  ]
| + command -v slirp4netns
| + [ -z  ]
| + command -v vpnkit
| + net=vpnkit
| + [ -z  ]
| + mtu=1500
| + [ -z  ]
| + _DOCKERD_ROOTLESS_CHILD=1
| + export _DOCKERD_ROOTLESS_CHILD
| + id -u
| + [ 0 = 0 ]
| + echo This script must be executed as a non-privileged user
| + exit 1
| This script must be executed as a non-privileged user
```

Reply:
Right, I don't think we can run this script as root (and not sure if it would be recommended anyway), but would have to try to be sure.

I think the issue here is that dockerd knows it is not real root, and while called from pseudo it thinks it is root because of the user id (which is 0 due pseudo). Maybe what we can do here (without changing docker) is trying to run this command as a user that is not root?

Reply:
The current implementation does not require running `dockerd` for preloading, it preloads just the skopeo store be means of `skopeo` utility. We just decided to go with non dockerd approach.

Reply:
It's been ready since the last week :). I am going to rebase it on the latest main, check whether it's bitbakable, and then merge.